src/api/c/canny.cpp

```diff
@@ -0,0 +1,248 @@
+}
+
+std::pair< Array<char>, Array<char> >
+computeCandidates(const Array<float>& supEdges, const float t1,
+                  const af_canny_threshold ct, const float t2)
+{
+    float maxVal = detail::reduce_all<af_max_t, float, float>(supEdges);
+    const unsigned NUM_BINS = static_cast<unsigned>(maxVal);
+
+    auto lowRatio = createValueArray<float>(supEdges.dims(), t1);
+
+    switch(ct)
+    {
+        case AF_AUTO_OTSU_THRESHOLD:
+            {
+                auto T2            = otsuThreshold(supEdges, NUM_BINS, maxVal);
+                auto T1            = arithOp<float, af_mul_t>(T2, lowRatio, T2.dims());
+                Array<char> weak1  = logicOp<float, af_ge_t >(supEdges,    T1, supEdges.dims());
+                Array<char> weak2  = logicOp<float, af_lt_t >(supEdges,    T2, supEdges.dims());
+                Array<char> weak   = logicOp<char , af_and_t>(   weak1, weak2,    weak1.dims());
+                Array<char> strong = logicOp<float, af_ge_t >(supEdges,    T2, supEdges.dims());
+                return std::make_pair(strong, weak);
+            };
+        default:
+            {
+                float minVal       = detail::reduce_all<af_min_t, float, float>(supEdges);
+                auto normG         = normalize(supEdges, minVal, maxVal);
+                auto T2            = createValueArray<float>(supEdges.dims(), t2);
+                auto T1            = createValueArray<float>(supEdges.dims(), t1);
+                Array<char> weak1  = logicOp<float, af_ge_t >(normG,    T1, normG.dims());
+                Array<char> weak2  = logicOp<float, af_lt_t >(normG,    T2, normG.dims());
+                Array<char> weak   = logicOp<char , af_and_t>(weak1, weak2, weak1.dims());
+                Array<char> strong = logicOp<float, af_ge_t >(normG,    T2, normG.dims());
+                return std::make_pair(strong, weak);
+            };
+    }
+}
+
+template<typename T>
+af_array cannyHelper(const Array<T> in, const float t1, const af_canny_threshold ct,
+                     const float t2, const unsigned sw, const bool isf)
+{
+    typedef typename std::pair< Array<float>, Array<float> > GradientPair;
+
+    static const std::vector<float> v = {-0.11021, -0.23691, -0.30576, -0.23691, -0.11021};
+    Array<float> cFilter= detail::createHostDataArray<float>(dim4(5, 1), v.data());
+    Array<float> rFilter= detail::createHostDataArray<float>(dim4(1, 5), v.data());
+
+    // Run separable convolution to smooth the input image
+    Array<float> smt = detail::convolve2<float, float, false>(cast<float, T>(in), cFilter, rFilter);
+
+    GradientPair g  = detail::sobelDerivatives<float, float>(smt, sw);
+    Array<float> gx = g.first;
+    Array<float> gy = g.second;
+
+    Array<float> gmag = gradientMagnitude(gx, gy, isf);
+
+    Array<float> supEdges = detail::nonMaximumSuppression(gmag, gx, gy);
+
+    auto swpair = computeCandidates(supEdges, t1, ct, t2);
```

Comment:
A side effect of working on it after a break. I will remove the typedef and qualify GradientPair declartion with auto.